# 第 2 章 简单回归模型 
## C2.1【简单回归；回归预测】  
### 401k.RAW 中的数据是帕普克（Papke，1995）所分析数据的一个子集。

帕普克是为了研究401 （k）养老金计划的参与率和该计划的慷慨程度之间的关系。变量prate 是有资格参与该计划 的员工中拥有活动账户的百分比，也是我们要解释的变量。慷慨程度指标是计划的匹配率 mrate。这个变量给出了员工每向这个账户存1美元，公司为该员工匹配的平均数量。例如mrate=0.50，则员工每投入1美元，公司就匹配50美分。 

（1）求出该计划的样本中平均参与率和平均匹配率。 

（2）现在估计下面这个简单回归方程
$\widehat{prate}=\beta_0+\beta_1 mrate$       报告你的结果以及样本容量和 $𝑅^2$。 

（3）解释你的方程中的截距。解释 mrate 的系数。 

（4）当mrate=3.5 时，求出prate 的预测值。这是一个合理的预测吗？解释这里出现的情况。 

（5）prate 的变异中，有多少是由 mrate 解释的？你认为，这是一个足够大的量吗？ 

# install.packages('haven')

In [2]:
library(haven)
#library(mice) #载入检查缺失值的包 mice。 
X401K <- read_dta("C:/Users/bradl/SRM/data/401K.DTA")
attach(X401K)
#md.pattern(X401K) 

The following objects are masked from X401K (pos = 3):

    age, ltotemp, mrate, prate, sole, totelg, totemp, totpart



（1）求出该计划的样本中平均参与率和平均匹配率。 

In [2]:
mean(prate) 
mean(mrate) 

[1] 87.36291

[1] 0.7315124

（2）现在估计下面这个简单回归方程
$\widehat{prate}=\beta_0+\beta_1 mrate$       报告你的结果以及样本容量和 $𝑅^2$。 

In [3]:
fit<-lm(prate~mrate)
summary(fit)


Call:
lm(formula = prate ~ mrate)

Residuals:
    Min      1Q  Median      3Q     Max 
-82.303  -8.184   5.178  12.712  16.807 

Coefficients:
            Estimate Std. Error t value Pr(>|t|)    
(Intercept)  83.0755     0.5633  147.48   <2e-16 ***
mrate         5.8611     0.5270   11.12   <2e-16 ***
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Residual standard error: 16.09 on 1532 degrees of freedom
Multiple R-squared:  0.0747,	Adjusted R-squared:  0.0741 
F-statistic: 123.7 on 1 and 1532 DF,  p-value: < 2.2e-16


In [14]:
length(residuals(fit))  
nobs(fit) 

[1] 1534

[1] 1534

（3）解释你的方程中的截距。解释 mrate 的系数。 

截距83.075 表示当mrate 为0 时，有83.075%的人参与养老金计划。mrate 的系数为5.861， 表示mrate 每提高一美元，prate 会增加5.861%。

（4）当mrate=3.5 时，求出prate 的预测值。这是一个合理的预测吗？解释这里出现的情况。 

In [4]:
new <- data.frame(mrate = 3.5) #生成新的数据框 new，内容为mrate=3.5 
predict(fit,new) 

1 
103.5892

（5）prate 的变异中，有多少是由 mrate 解释的？你认为，这是一个足够大的量吗？

R-squared:  0.0747，所以有7.47%是由mrate解释的。这并不算大。因此还可能有其他变量会对prate的变异进行解释，但我们并没有纳入我们的模型中。 

In [5]:
detach(X401K)

## C2.2【因变量为 log(y)的回归；频数表】  
### 数据集CEOSAL2.RAW 中包含了美国公司首席执行官的信息。变量salary 是以千美元计的年薪，ceoten 是已担任公司CEO的年数。 

（1）求出样本中的平均年薪和平均任期。 

（2）有多少CEO 尚处于担任CEO 的第一年（也就是说，ceoten=0）?最长的CEO 任期是多 少？ 

（3）估计简单回归模型 $\log(salary)=\beta_0+\beta_1 ceoten + u $,用通常的形式报告你的结果。多担任一年CEO，预计年薪增长（近似）的百分数是多少？ 

In [8]:
CEOSAL2 <- read_dta("C:/Users/bradl/SRM/data/CEOSAL2.DTA")
attach(CEOSAL2)
#md.pattern(CEOSAL2) 

（1）求出样本中的平均年薪和平均任期。 

In [9]:
mean(salary)
mean(ceoten) 

[1] 865.8644

[1] 7.954802

（2）有多少CEO 尚处于担任CEO 的第一年（也就是说，ceoten=0）?最长的CEO任期是多少？ 

In [11]:
table(ceoten) #获得CEO 任期变量 ceoten 的频数表： 
ceoten 

ceoten
 0  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20 21 22 24 26 28 
 5 19 10 21 21 10 11  6 11  8  8  4  7  7  5  2  2  2  1  2  4  1  1  3  2  1 
34 37 
 1  2 

[1]  2 10  3 22  6  7 10  8  4  5  7  8 37  1 11  7  4 12 24  1  6 14  4  8  4
 [26]  4  6  2  8 26  4  9  1 14  1 13  4  4  5 24 13  8  3  3  3  8  8  1  3  1
 [51]  2  7  4 20  5 12 19  9 13  8 15 21  1  4  7  2  1  4 17 10  6 10 13 11  4
 [76]  6  6  5  9 12 14  3  3  5  0 13  4  7  3 10  5 10  1  4  3  3 12  4  1 10
[101] 11  2 13  3  3  3  8  5  1 15  1 24 26  6 10  2  6 14  6  4 37 28  4  1  4
[126] 20  5 19  0  8  1  9 14  8 20 34  0  9 16  3  1  9  3  3 12  9  6  4  3  9
[151]  6  4  2  1  3 20 12 12 11 17  2  1  2  5  4  3  3  1  2  0 16  5  3  1 13
[176] 18  0
attr(,"label")
[1] "years as ceo with company"
attr(,"format.stata")
[1] "%9.0g"

（3）估计简单回归模型 $\log(salary)=\beta_0+\beta_1 ceoten + u $,用通常的形式报告你的结果。多担任一年CEO，预计年薪增长（近似）的百分数是多少？ 

In [12]:
fit<-lm(log(salary)~ceoten) #做简单回归，把结果保存在fit 中。 
summary(fit) 


Call:
lm(formula = log(salary) ~ ceoten)

Residuals:
     Min       1Q   Median       3Q      Max 
-2.15314 -0.38319 -0.02251  0.44439  1.94337 

Coefficients:
            Estimate Std. Error t value Pr(>|t|)    
(Intercept) 6.505498   0.067991  95.682   <2e-16 ***
ceoten      0.009724   0.006364   1.528    0.128    
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Residual standard error: 0.6038 on 175 degrees of freedom
Multiple R-squared:  0.01316,	Adjusted R-squared:  0.007523 
F-statistic: 2.334 on 1 and 175 DF,  p-value: 0.1284


In [13]:
detach(CEOSAL2)

## C2.3【简单回归】  
### 利用Biddle and Hamermesh (1990)中的SLEEP75.RAW 数据，研究在每周用于睡眠的时间和用于有酬工作的时间之间是否存在替代关系。

我们可以用它们中的任何一个作为因变量。为具体起见，估计模型 $sleep=\beta_0+\beta_1 totwrk+u$ 其中，sleep 是每周用于晚上睡眠的分钟数，totwrk是这一周中用于工作的分钟数。 

（1）用方程的形式，连同观测的次数和$𝑅^2$报告你的结果。该方程中的截距表示什么？ 

（2）若totwrk 增加2小时，则sleep估计要减少多少？你觉得这是一个很大的效应吗？ 

In [14]:
SLEEP75 <- read_dta("C:/Users/bradl/SRM/data/SLEEP75.DTA")
attach(SLEEP75)

The following object is masked from package:datasets:

    sleep



（1）用方程的形式，连同观测的次数和$𝑅^2$报告你的结果。该方程中的截距表示什么？ 


In [15]:
fit<-lm(sleep~totwrk) #对模型做回归 
summary(fit) 


Call:
lm(formula = sleep ~ totwrk)

Residuals:
     Min       1Q   Median       3Q      Max 
-2429.94  -240.25     4.91   250.53  1339.72 

Coefficients:
              Estimate Std. Error t value Pr(>|t|)    
(Intercept) 3586.37695   38.91243  92.165   <2e-16 ***
totwrk        -0.15075    0.01674  -9.005   <2e-16 ***
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Residual standard error: 421.1 on 704 degrees of freedom
Multiple R-squared:  0.1033,	Adjusted R-squared:  0.102 
F-statistic: 81.09 on 1 and 704 DF,  p-value: < 2.2e-16


In [16]:
length(residuals(fit)) 
nobs(fit) 

[1] 706

[1] 706

（2）若totwrk 增加2 小时，则sleep 估计要减少多少？你觉得这是一个很大的效应吗？ 

totwrk 增加 2 个小时则sleep 减少约 0.3 个小时。这个效应不算大。 

In [17]:
detach(SLEEP75)  

## C2.4【因变量为log(y)的回归】
### 利用WAGE2.RAW中的数据估计一个简单回归，以便用智商（IQ）来解释月薪（wage）。

（1）求出样本中的平均工资和平均IQ。IQ的样本标准差是多少？（总体中的IQ已标准化为平均值是100，标准差是15。）

（2）估计一个简单回归模型，其中IQ提高一个单位导致wage变化相同的数量。利用这个模型计算IQ提高15个单位时，工资的预期变化。IQ能够解释大多数工资变异吗？

（3）现在再估计一个模型，其中IQ提高一个单位对工资具有相同的百分比影响。如果IQ提高15个单位，预期工资提高的百分比大约是多少？

In [19]:
data(wage2)
WAGE2<-wage2
attach(WAGE2)
#library(mice) #载入检查缺失值的包mice。
#md.pattern(WAGE2) #

The following objects are masked from WAGE2 (pos = 3):

    age, black, brthord, educ, exper, feduc, hours, IQ, KWW, lwage,
    married, meduc, sibs, south, tenure, urban, wage



（1）求出样本中的平均工资和平均IQ。IQ的样本标准差是多少？

In [20]:
mean(wage) #平均工资
mean(IQ) #平均IQ
sd(IQ) 

[1] 957.9455

[1] 101.2824

[1] 15.05264

（2）估计一个简单回归模型，其中IQ提高一个单位导致wage变化相同的数量。利用这个模型计算IQ提高15个单位时，工资的预期变化。IQ能够解释大多数工资变异吗?
设定模型 $wage=\beta_0+\beta_1 IQ+\mu$

In [21]:
fit1<-lm(wage~IQ) #估计上述模型
summary(fit1) #


Call:
lm(formula = wage ~ IQ)

Residuals:
   Min     1Q Median     3Q    Max 
-898.7 -256.5  -47.3  201.1 2072.6 

Coefficients:
            Estimate Std. Error t value Pr(>|t|)    
(Intercept) 116.9916    85.6415   1.366    0.172    
IQ            8.3031     0.8364   9.927   <2e-16 ***
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Residual standard error: 384.8 on 933 degrees of freedom
Multiple R-squared:  0.09554,	Adjusted R-squared:  0.09457 
F-statistic: 98.55 on 1 and 933 DF,  p-value: < 2.2e-16


（3）现在再估计一个模型，其中IQ提高一个单位对工资具有相同的百分比影响。如果IQ提高15个单位，预期工资提高的百分比大约是多少?
设定模型$\log(wage)=\beta_0+\beta_1 IQ+\mu$

In [22]:
fit2<-lm(log(wage)~IQ) #估计上述模型
summary(fit2) #


Call:
lm(formula = log(wage) ~ IQ)

Residuals:
     Min       1Q   Median       3Q      Max 
-2.09324 -0.25547  0.02261  0.27544  1.21486 

Coefficients:
             Estimate Std. Error t value Pr(>|t|)    
(Intercept) 5.8869943  0.0890206   66.13   <2e-16 ***
IQ          0.0088072  0.0008694   10.13   <2e-16 ***
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Residual standard error: 0.3999 on 933 degrees of freedom
Multiple R-squared:  0.09909,	Adjusted R-squared:  0.09813 
F-statistic: 102.6 on 1 and 933 DF,  p-value: < 2.2e-16


In [44]:
detach(WAGE2)

## C2.5【常弹性模型】
### 在化工产业的企业总体中，令rd表示年研发支出，sales表示年销售额（都以百万美元计）。

（1）写一个模型（不是估计方程），其中rd和sales之间的弹性为常数。哪一个参数代表弹性？
常弹性模型：$\log{rd}=\beta_0+\beta_1\log{sales}+u$，其中$\beta_1$表示弹性。

（2）再用RDCHEM.RAW中的数据估计模型。用通常的形式写出估计方程。rd关于sales的弹性的估计值是多少？用文字解释这个弹性的含义。

In [23]:
data(rdchem)
RDCHEM<-rdchem
attach(RDCHEM)
#library(mice) #载入检查缺失值的包mice。
#md.pattern(RDCHEM) #

（1）写一个模型（不是估计方程），其中rd和sales之间的弹性为常数。哪一个参数代表弹性？
常弹性模型：$\log{rd}=\beta_0+\beta_1\log{sales}+u$，其中$\beta_1$表示弹性。

In [24]:
fit<-lm(log(rd)~log(sales)) #做（1）中常弹性模型的回归。
summary(fit)


Call:
lm(formula = log(rd) ~ log(sales))

Residuals:
     Min       1Q   Median       3Q      Max 
-0.90406 -0.40086 -0.02178  0.40562  1.10438 

Coefficients:
            Estimate Std. Error t value Pr(>|t|)    
(Intercept) -4.10472    0.45277  -9.066 4.27e-10 ***
log(sales)   1.07573    0.06183  17.399  < 2e-16 ***
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Residual standard error: 0.5294 on 30 degrees of freedom
Multiple R-squared:  0.9098,	Adjusted R-squared:  0.9068 
F-statistic: 302.7 on 1 and 30 DF,  p-value: < 2.2e-16


In [25]:
length(residuals(fit))
detach(RDCHEM)

[1] 32

（2）再用RDCHEM.RAW中的数据估计模型。用通常的形式写出估计方程。rd关于sales的弹性的估计值是多少？用文字解释这个弹性的含义。

## C2.6【自变量为log(x)的回归；回归预测】
### 例2.12中曾使用了MEAP93.RAM中的数据。现在，我们想用这个文件中的数据来说明数学通过率（math10）与没个学生的平均支出（expend）之间的关系。

（1）就多花一美元对通过率的影响而言，你认为具有恒定不变的影响合适呢，还是这种影响越来越小合适？请加以解释。

答：越来越小合适。对于低支出的学生来说，1美元支出带来的边际效用理应更大，而随着边际效用递减，对于高支出的学生来说，1美元支出带来的效用更低。

（2）在总体模型 $math10=\beta_0+\beta_1\log{expend}+u$中，证明$\beta_1/10$表示expend提高了10\%导致math10改变的百分数。

答：该式表示expend提高1%，math10将提高$\beta_1/100$个单位。所以$\beta_1/10$表示expend提高了10%导致math10改变的单位数，也就是通过率增加的百分数。

（3）利用MEAP93.RAM中的数据，估计（2）中的模型。按照通常的方式报告估计方程，包括样本容量和$R^2$。

（4）支出的估计影响有多大？也就是说，如果支出提高10\%，估计math10会提高多少个百分点？

（5）有人担心这个回归分析可能达到math10的拟合值会超过100。为什么在这个数据集中不必担心这个问题？

In [56]:
data(meap93)
MEAP93<-meap93
attach(MEAP93)

The following object is masked from package:wooldridge:

    benefits



（3）利用MEAP93.RAM中的数据，估计（2）中的模型。按照通常的方式报告估计方程，包括样本容量和$R^2$。

In [57]:
fit<-lm(math10~log(expend)) #做（2）中模型的回归。
summary(fit) #
nobs(fit)


Call:
lm(formula = math10 ~ log(expend))

Residuals:
    Min      1Q  Median      3Q     Max 
-22.343  -7.100  -0.914   6.148  39.093 

Coefficients:
            Estimate Std. Error t value Pr(>|t|)    
(Intercept)  -69.341     26.530  -2.614 0.009290 ** 
log(expend)   11.164      3.169   3.523 0.000475 ***
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Residual standard error: 10.35 on 406 degrees of freedom
Multiple R-squared:  0.02966,	Adjusted R-squared:  0.02727 
F-statistic: 12.41 on 1 and 406 DF,  p-value: 0.0004752


[1] 408

（4）支出的估计影响有多大？也就是说，如果支出提高10\%，估计math10会提高多少个百分点？
11.164/10=1.1164

（5）有人担心这个回归分析可能达到math10的拟合值会超过100。为什么在这个数据集中不必担心这个问题?

In [58]:
max(expend) #求在数据集中log(expend)的最大值为多少，结果是7419。
new <- data.frame(expend = max(expend)) #生成新的数据框new，内容为mrate中的最大值7419。
predict(fit,new) #用predict()函数进行预测，得到结果30.15376，远小于100。所以在这个数据集中，即便expend取最大值，math10距离100也还差很远。
detach(MEAP93) #

[1] 7419

1 
30.15376

## C2.7【按条件选入观测；简单回归；工整导出回归结果】
### 利用CHARITY.RAW中的数据[得自于Franses and Paap (2001)]回答如下问题：

（1）在这个4268人的样本中，平均捐款数量是多少（以荷兰盾为单位）？没有捐款的人数百分比是多少？

（2）每年平均寄出的邮件数量是多少？其最小值和最大值是多少？

（3）用普通最小二乘法估计如下模型： $gift=\beta_0+\beta_1 mailsyear+u$ 按照通常的方式报告估计方程，包括样本容量和$R^2$。

（4）解释斜率系数。如果每封邮件的成本是1盾，那么慈善机构预期能够从寄出的每一封邮件中获得净利润吗？这意味着慈善机构从每封邮件中都获得了净利润吗？请加以解释。

（5）样本中最小慈善捐款的预测值是多少？利用这个简单的回归分析，你有可能预测gift等于0吗？

In [61]:
data(charity)
attach(charity)

In [62]:
mean(gift) 
a<-charity[gift==0,] #将gift为0的观测挑选出来保存在a中。
length(a$gift) 
length(a$gift)/length(charity$gift) #

[1] 7.44447

[1] 2561

[1] 0.6000469

In [63]:
mean(mailsyear) 
min(mailsyear) 
max(mailsyear)#

[1] 2.049555

[1] 0.25

[1] 3.5

In [65]:
fit<-lm(gift~mailsyear) #估计上述模型。
library(stargazer) 
stargazer(fit, type="text") # 工


Please cite as: 

 Hlavac, Marek (2018). stargazer: Well-Formatted Regression and Summary Statistics Tables.
 R package version 5.2.2. https://CRAN.R-project.org/package=stargazer 




                        Dependent variable:    
                    ---------------------------
                               gift            
-----------------------------------------------
mailsyear                    2.650***          
                              (0.343)          
                                               
Constant                     2.014***          
                              (0.739)          
                                               
-----------------------------------------------
Observations                   4,268           
R2                             0.014           
Adjusted R2                    0.014           
Residual Std. Error     14.960 (df = 4266)     
F Statistic          59.649*** (df = 1; 4266)  
Note:               *p<0.1; **p<0.05; ***p<0.01


# C:\Program Files\R\R-3.6.1\library